spec/API_specification/data_type_functions.md

```diff
@@ -0,0 +1,84 @@
+# Data type functions
+
+> Array API specification for data type functions.
+
+A conforming implementation of the array API standard must provide and support the following data type functions.
+
+<!-- NOTE: please keep the constants in alphabetical order -->
+
+## Objects in API
+
+(finfo)=
+### finfo(type)
+
+Defines the machine limits for floating point types.
+
+#### Parameters
+
+-   **type**: _Union\[ &lt;dtype&gt, instance ]_
+
+    -   the kind of floating point data-type about which to get information
+
+#### Returns
+
+-   **out**: _&lt;class&gt;_
+
+    -   a class with that encapsules the following attributes:
+
+        -   **bits**: _int_
+            -   The number of bits occupied by the type
+        -   **eps**: _float_
+            -   The difference between 1.0 and the next smallest representable float larger than 1.0 following the IEEE 754 standard.
+        -   **max**: _float_
+            -   The largest representable number.
+        -   **min**: _float_
+            -   The smallest representable number.
+        -   **tiny**: _float_
```

Comment:
@kgryte do you happen to know if this is guaranteed to be the same across implementations? I.e. it's part of IEEE 754 somehow?

Reply:
This is what the NumPy docs have in their notes:
> Note that tiny is not actually the smallest positive representable value in a NumPy floating point type. As in the IEEE-754 standard [1], NumPy floating point types make use of subnormal numbers to fill the gap between 0 and tiny. However, subnormal numbers may have significantly reduced precision [2].

Reply:
No. While part of IEEE 754, support for subnormal numbers is less common, especially among older generation GPUs. Often subnormal numbers trigger slow paths, and, if not implemented in hardware, software emulation. More recent NVIDIA GPUs do support subnormal numbers. And among ARM processors, one can flush subnormal numbers to zero.

Based on NumPy docs (via @steff456 ), `tiny` refers to the smallest normal number. We should probably be explicit by what we mean by "representable number".

Reply:
Interesting. So the description is a bit off here. I'm not sure it should be part of the standard. I can imagine it may be useful in a few corner cases, but I've never seen it used in the wild.

Let's see if anyone else has an opinion.

Reply:
Okay, so there are use cases at the compiled code level at least.

Reply:
I suppose having this info is useful for meta inspection concerning platform capabilities. For example, if I want to know whether a platform supports subnormals. Apart from meta data, a user would need to compute a quantity which should resolve to a subnormal number and manually check whether the value is zero.

Reply:
Correct. It can also come up in user code, e.g., in the computation of transcendentals, where, if you know that a platform does not support subnormals, then you can avoid various, often slower, branching logic.

Reply:
I think if we wanted to be complete, instead of "tiny", we'd have `smallest_normal` and `smallest_subnormal`. If the latter is `0`, then a platform does not support subnormal numbers.

Reply:
Those do seem like better names indeed. TF and CuPy just alias `numpy.finfo`, which may be wrong. And JAX implements it but docs say _"The smallest positive usable number"_, which is quite uninformative.

I think we can add the new names, but then we should make a PR to NumPy to see if that's accepted. And otherwise we should probably drop it completely.